liquid-rust-syntax/src/surface.rs

```diff
@@ -0,0 +1,209 @@
+    pub ensures: Vec<(Ident, Ty)>,
+    /// source span
+    pub span: Span,
+}
+
+#[derive(Debug)]
+pub struct Ty {
+    pub kind: TyKind,
+    pub span: Span,
+}
+
+#[derive(Debug)]
+pub enum TyKind {
+    /// ty
+    Base(Path),
+
+    /// ty{b:e}
+    Exists { bind: Ident, path: Path, pred: Expr },
+
+    /// ty{e}, the param binder is used e.g. x: i32{0 < x}
+    AnonEx { path: Path, pred: Expr },
+
+    /// named: n@t
+    Named(Ident, Box<Ty>),
+
+    /// reference
+    Ref(RefKind, Box<Ty>),
+}
+
+#[derive(Debug)]
+pub struct Path {
+    /// vec
+    pub ident: Ident,
+    /// <nat>
+    pub args: Option<Vec<Ty>>,
+    pub span: Span,
+}
+
+// #[derive(Debug)]
+// pub struct NamedTy {
+//     pub kind: NamedTyKind,
+//     pub span: Span,
+// }
+
+// #[derive(Debug)]
+// pub enum NamedTyKind {
+//     /// For inputs
+//     NamedBase(Ident, Ty),
+
+//     /// For outputs
+//     AnonBase(Ty),
+
+//     /// For inputs and outputs
+//     Ref(RefKind, Box<NamedTy>),
+// }
+
+#[derive(Debug)]
+pub enum RefKind {
+    Mut,
+    Immut,
+}
+
+struct Desugar {
+    generics: Vec<ast::GenericParam>,
+    args: Vec<ast::Ty>,
+    reqs: Vec<(Ident, ast::Ty)>,
+}
+
+/// A `BindIn` is the information obtained from a single input-param binding
+struct BindIn {
+    gen: Option<GenericParam>,
+    ty: ast::Ty,
+    loc: Option<(Ident, ast::Ty)>,
+}
+
+fn convert_path(p: Path) -> ast::Path {
+    ast::Path {
+        ident: p.ident,
+        span: p.span,
+        args: match p.args {
+            Some(ts) => Some(ts.into_iter().map(convert_ty).collect()),
+            None => None,
+        },
+    }
+}
+
+fn convert_tykind(t: TyKind) -> ast::TyKind {
+    match t {
+        TyKind::Base(p) => ast::TyKind::BaseTy(convert_path(p)),
+        TyKind::Exists { bind, path, pred } => {
+            ast::TyKind::Exists { bind, path: convert_path(path), pred }
+        }
+        TyKind::AnonEx { .. } => {
+            panic!("Unexpected input in convert_tykind!")
+        }
+        TyKind::Named(_, t) => {
+            let ty = convert_ty(*t);
+            ty.kind
+        }
+        TyKind::Ref(_, t) => ast::TyKind::WeakRef(Box::new(convert_ty(*t))),
+    }
+}
+
+fn convert_ty(t: Ty) -> ast::Ty {
+    ast::Ty { kind: convert_tykind(t.kind), span: t.span }
+}
+
+fn mk_sort(span: Span) -> Ident {
+    Ident { name: rustc_span::Symbol::intern("int"), span }
+}
+
+fn mk_singleton(x: Ident) -> ast::Refine {
+    let e = Expr { kind: ast::ExprKind::Var(x), span: x.span };
+    ast::Refine { exprs: vec![e], span: x.span }
+}
+
+fn mk_generic(x: Ident, pred: Option<Expr>) -> ast::GenericParam {
+    ast::GenericParam { name: x, sort: mk_sort(x.span), pred }
+}
+
+fn ident_loc(x: Ident) -> Ident {
+    let mut star = String::from("*");
+    star.push_str(x.name.as_str());
+    let name = rustc_span::Symbol::intern(&star);
+    Ident { name, span: x.span }
+}
+
+fn path_bind_in(x: Ident, path: Path, span: Span, pred: Option<Expr>) -> BindIn {
```

Comment:
You'd typically have these "constructors" as an associated function in the corresponding type

```rust
impl BindIn {
    fn from_path(x: Ident, path: Path, span: Span, pred: Option<Expr>) -> BindIn {
        ...
    }
}
```